edx_proctoring/static/index.js

```diff
@@ -0,0 +1,32 @@
+export const handlerWrapper = (Handler) => {
+   let handler = new Handler({});
+
+  self.addEventListener("message", (message) => {
+    switch(message.data.type) {
+      case 'config': {
+        handler = new Handler(message.data.options);
+        break;
+      }
+      case 'startExamAttempt': {
+        if(handler.onStartExamAttempt) {
+          handler.onStartExamAttempt().then(() => self.postMessage({type: 'examAttemptStarted'}))
+        }
+        break;
+      }
+      case 'endExamAttempt': {
+        if(handler.onEndExamAttempt) {
+          handler.onEndExamAttempt().then(() => self.postMessage({type: 'examAttemptEnded'}))
+        }
+        break;
+      }
+      case 'ping': {
+        if(handler.onPing) {
+          handler.onPing().then(() => self.postMessage({type: 'echo'}))
```

Comment:
Doesn't seem to be used in the action handler at all, but you may want to be consistent and call this type 'ping'

Reply:
I pushed it off, but the implementation is in progress: https://github.com/edx/edx-proctoring/pull/466

When this was first pitched to me it was described as a "heartbeat" mechanism, and I found the ping/echo language more descriptive when googling around to understand the rationale for implementing the pattern: http://etutorials.org/Programming/Software+architecture+in+practice,+second+edition/Part+Two+Creating+an+Architecture/Chapter+5.+Achieving+Qualities/5.2+Availability+Tactics/

Reply:
Agreed the ping/echo makes sense, but it seems the type of message would be `ping` while the message itself, if we were sending data along, would be `echo`, I was just getting hung up on the `onPing()` responding by sending an `echo` type